doc/design/api.md

```diff
@@ -0,0 +1,87 @@
+1. *updater*, which encapsulates the updating algorithm.
+   
+It seems that *cost function* is a property of *gradient machine*?
+   
+### Data
+
+Models are trained using data sets.  We hope to provide a set of utility data sets encapsulated in Python packages like `paddle.data.amazon.product_review` and `paddle.data.wikipedia.articles`.  A reasonable idea might be that in each of these packages, we provide a `new` function that returns a reader object or a Python iterator.  And the *reader* has a read method `read`, which, once called, returns a minibatch and a flag indicating if it reaches the end of a data set.  For online learning, the flag would always be False.  In this way, we don't have to have two concepts: *pass* and *iteration*; instead, we need only the latter.
+
+## Examples
+
+### A Simple Network
+
+```python
+gm = paddle.gradient_machine.new(model)  # gm uses default cost function.
+ud = paddle.updater.new_simple()  # A simple updater doesn't work with parameter servers.
+rd = paddle.data.amazon.product_review.new()
+mt = paddle.metric.new_acc()
+
+for mb, pass_end in rd.read():
+    gm.feed(mb)
+    ud.update(gm)
+    mt.append(gm, mb) # compute and record the accuracy of this minibatch.
+    if pass_end:
+        log(mt.flash()) # output aggregated accuracy records of this pass and reset mt.
+```
+
+In this example, `GradientMachine.feed` is a convenience that calls `GradientMachine.forward` and `GradientMachine.backward`.
+
+### A GAN Example
+
+```python
+input_gen = paddle.layer.input(...)
+hidden_gen = paddle.layer.fc(input_gen, ...)
+output_gen = paddle.layer.fc(hidden_gen, ...)
+
+# gm_gen maintains layer outputs and gradients of model gen.
+gm_gen = paddle.gradient_machine.new(output_gen)
+
+input_dis = paddle.layer.input(...)
+hidden_dis = paddle.layer.fc(intput_dis, ...)
+output_dis = paddle.layer.softmax(hidden_dis, ...)
+
+# gm_dis maintains layer outputs and gradients of model dis.
+gm_dis = paddle.gradient_machine.new(output_dis)
+
+# gm_comp maintains layer outputs and gradients of both gen and dis.
+gm_comp = paddle.gradient_machine.compose(output_gen, output_dis)
+
+ud = paddle.updater.new_simple()
+
+rd = paddle.data.mnist.new()
+
+for mb, pass_end in rd.read():
+    fake = gm_gen.forward(mb)
+    fake_label = paddle.input.new(False, ...)
+    real_label = paddle.input.new(True, ...)
+    gm_dis.feed([fake, fake_label])
+    gm_dis.feed([mb, real_label])
+    ud.update(gm_dis)
+    
+    gm_comp.feed([mb, real_label])
+    ud.update(gm_comp, output_gen) # updates only the model whose output layer is output_gen.
+```
+
+A key point here is that we use the output layer to indicate a model.  I think that we can achieve this as long as each layer knows about its predecessors so that we can trace from the output layer upward till the input layers.  Please be aware that we didn't compose two models in above example code; instead, we only created a gradient machine which covers both `model_gen` and `model_dis`.
```

Comment:
Just for reference: If I understand correctly, tensorflow do it by explicitly state weights that needs update.
```
fine_tune_step = tf.train.AdamOptimizer(1e-2).minimize(cross_entropy, var_list=[W_fc2, b_fc2])
fine_tune_step.run(feed_dict={x: batch[0], y_: batch[1], keep_prob: 0.5})
```